libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.kt

```diff
@@ -0,0 +1,142 @@
+     *
+     *
+     * If the coordinate size is less than the remaining waypoints, this method
+     * will return null.
+     *
+     * @param routeProgress for route coordinates and remaining waypoints
+     * @return list of remaining waypoints as [Point]s
+     * @since 0.10.0
+     */
+    fun calculateRemainingWaypoints(routeProgress: RouteProgress): List<Point>? {
+        val routeOptions = routeProgress.directionsRoute().routeOptions() ?: return null
+        val coordinates = ArrayList(routeOptions.coordinates())
+        val coordinatesSize = coordinates.size
+        val remainingWaypoints = routeProgress.remainingWaypoints()
+        return when (coordinatesSize < remainingWaypoints) {
+            true -> null
+            false -> coordinates.subList(coordinatesSize - remainingWaypoints, coordinatesSize)
+        }
+    }
+
+    /**
+     * Given a [RouteProgress], this method will calculate the remaining waypoint names
+     * along the given route based on route option waypoint names and the progress remaining coordinates.
+     *
+     *
+     * If the waypoint names are empty, this method will return null.
+     *
+     * @param routeProgress for route waypoint names and remaining coordinates
+     * @return String array including the origin waypoint name and the remaining ones
+     * @since 0.19.0
+     */
+    fun calculateRemainingWaypointNames(routeProgress: RouteProgress): Array<String?>? {
+        val routeOptions = routeProgress.directionsRoute().routeOptions() ?: return null
+        val allWaypointNames = routeOptions.waypointNames()
+        if (allWaypointNames == null || TextUtils.isEmpty(allWaypointNames)) {
+            return null
+        }
+        val names = allWaypointNames.split(SEMICOLON.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
+        val coordinatesSize = routeOptions.coordinates().size
+        val remainingWaypointNames = Arrays.copyOfRange(names,
+                coordinatesSize - routeProgress.remainingWaypoints(), coordinatesSize)
+        val waypointNames = arrayOfNulls<String>(remainingWaypointNames.size + ORIGIN_WAYPOINT_NAME_THRESHOLD)
+        waypointNames[ORIGIN_WAYPOINT_NAME] = names[ORIGIN_WAYPOINT_NAME]
+        System.arraycopy(remainingWaypointNames, FIRST_POSITION, waypointNames, SECOND_POSITION,
+                remainingWaypointNames.size)
+        return waypointNames
+    }
+
+    /**
+     * Given the current step / current step distance remaining, this function will
+     * find the current instructions to be shown.
+     *
+     * @param currentStep holding the current banner instructions
+     * @param stepDistanceRemaining to determine progress along the currentStep
+     * @return the current banner instructions based on the current distance along the step
+     * @since 0.13.0
+     */
+    fun findCurrentBannerInstructions(currentStep: LegStep?, stepDistanceRemaining: Double): BannerInstructions? {
```

Comment:
This is only used in `InstructionListPresenter`, should we considering removing it and adapt UI module?